### PR TITLE
Remote source indexing

### LINKED ISSUE #2191: CompositeView: Implement indexing logic for remote sources

### PR DESCRIPTION
**Pull request by wwajerowicz**
Fixes #2191

### CODE REVIEW

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala [hunk 1]
```diff
@@ -31,15 +34,15 @@ final case class BlazegraphIndexingStreamEntry(
...
 
   /**
     * Generates an Sparql replace query with all the triples to be added to the resource named graph
     */
   def index(includeMetadata: Boolean): Task[Option[SparqlWriteQuery]] =
     for {
-      triples    <- toTriples(includeMetadata)
-      namedGraph <- namedGraph(resource.resolvedId)
+      triples    <- IO.fromEither(toTriples(includeMetadata))
+      namedGraph <- namedGraph(resource.id)
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala:45]: Why this is reverted? To use `resource.resolvedId` was a fix for projects.

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala [hunk 2]
```diff
@@ -31,15 +34,15 @@ final case class BlazegraphIndexingStreamEntry(
     * Deletes the current resource named graph (and all its triples)
     */
   def delete(): Task[SparqlWriteQuery] =
-    namedGraph(resource.resolvedId).map(SparqlWriteQuery.drop)
+    namedGraph(resource.id).map(SparqlWriteQuery.drop)
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala:37]: Why this is reverted? To use `resource.resolvedId` was a fix for projects.

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala [hunk 3]
```diff
@@ -87,11 +92,52 @@ object BlazegraphIndexingStreamEntry {
...
+        .map(_._3)
+        .toRight(MissingPredicate(predicate))
+    }
+
+    val idNode = Triple.subject(id)
+    for {
+      graph      <- Graph(nQuads)
+      valuegraph  = graph.filter { case (_, p, _) => !metadataPredicates.contains(p) }
+      metagraph   = graph.filter { case (_, p, _) => metadataPredicates.contains(p) }
+      types       = graph
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala:130]: We have a method in `Graph` for this. `graph.rootTypes`

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala [hunk 4]
```diff
@@ -87,11 +92,52 @@ object BlazegraphIndexingStreamEntry {
...
+  /**
+    * Converts the resource in n-quads fromat to [[BlazegraphIndexingStreamEntry]]
+    */
+  def fromNQuads(
+      id: Iri,
+      nQuads: NQuads,
+      metadataPredicates: Set[Node]
+  ): Either[RdfError, BlazegraphIndexingStreamEntry] = {
+
+    def findPredicate(idNode: Node, graph: Graph, predicate: Iri): Either[RdfError, Node] = {
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala:114]: It seems that what you want is something in `Graph` class like this:
```scala
def find(s: IriOrBNode, p: Iri): Option[Node] = ???
```

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala [hunk 5]
```diff
@@ -87,11 +92,52 @@ object BlazegraphIndexingStreamEntry {
...
     } yield BlazegraphIndexingStreamEntry(data)
   }
+
+  /**
+    * Converts the resource in n-quads fromat to [[BlazegraphIndexingStreamEntry]]
+    */
+  def fromNQuads(
+      id: Iri,
+      nQuads: NQuads,
+      metadataPredicates: Set[Node]
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala:111]: I would rather have an own case class for this, not just a plain St,

```scala
final case class MetadataPredicates(values: Set[Node])
```

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeViewsPluginModule.scala [hunk 6]
```diff
@@ -108,6 +109,12 @@ class CompositeViewsPluginModule(priority: Int) extends ModuleDef {
     new IndexingStreamController[CompositeView](CompositeViews.moduleType)(as)
   }
 
+  make[JsonLdContext].fromEffect {
```
Comment (umbreak) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeViewsPluginModule.scala:112]:
```scala
make[MetadataPredicates].fromEffect {
(aggMetadataCtx: MetadataContextValue @Id("aggregated-metadata"), cr: RemoteContextResolution @Id("aggregate")) =>
      implicit val res = cr
      val jsonLdContext = JsonLdContext(aggMetadataCtx.value)
      MetadataPredicates(jsonLdContext.aliasesInv.keySet.map(Triple.predicate))
}
```

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeViewsPluginModule.scala [hunk 7]
```diff
@@ -132,7 +140,8 @@ class CompositeViewsPluginModule(priority: Int) extends ModuleDef {
         projectsCounts,
         indexingController,
         projection,
-        indexingSource
+        indexingSource,
+        metadataContext.aliasesInv.keySet.map(Triple.predicate)
```
Comment (umbreak) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeViewsPluginModule.scala:144]: You don't need that anymore

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/client/RemoteSse.scala [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.client
+
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import io.circe.Decoder
+import io.circe.generic.extras.Configuration
+import io.circe.generic.extras.semiauto.deriveConfiguredDecoder
+
+import java.time.Instant
+
+/**
```
Comment (umbreak) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/client/RemoteSse.scala:10]: formatting alignment

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingStream.scala [hunk 9]
```diff
@@ -60,9 +64,13 @@ final class CompositeIndexingStream(
...
     projectsCounts: ProjectsCounts,
     remoteProjectsCounts: RemoteProjectsCounts,
+    remoteProjectStream: RemoteProjectStream,
+    remoteResourceNQuads: RemoteResourceNQuads,
+    remoteSourceConfig: RemoteSourceClientConfig,
     restartProjections: RestartProjections,
     projections: Projection[Unit],
-    indexingSource: IndexingSource
+    indexingSource: IndexingSource,
+    metadataPredicates: Set[Node]
```
Comment (umbreak) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingStream.scala:73]: MetadataPredicates

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingStream.scala [hunk 10]
```diff
@@ -96,20 +104,41 @@ final class CompositeIndexingStream(
...
+            case s: ProjectSource       =>
+              indexingSource(view.projectRef, progress.offset, s.resourceTag).evalMapValue(
+                BlazegraphIndexingStreamEntry.fromEventExchange(_)
+              )
+            case s: CrossProjectSource  =>
+              indexingSource(s.project, progress.offset, s.resourceTag).evalMapValue(
+                BlazegraphIndexingStreamEntry.fromEventExchange(_)
+              )
+            case s: RemoteProjectSource =>
+              remoteProjectStream(s, progress.offset)
```
Comment (umbreak) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingStream.scala:120]: All this block seems to much here. It was already complicated, now is even more.
All this part, for `ProjectSource` and `CrossProjectSource` is handled in the `IndexingSource`

It should be the same for remote projects. 

The easier way would be to add `CompositeViewIndexingSource` as follows:
```scala
trait CompositeViewIndexingSource {
  def apply(
      project: ProjectRef,
      offset: Offset,
      tag: Option[TagLabel]
  ): Stream[Task, Chunk[Message[BlazegraphIndexingStreamEntry]]]
}
```

And then on the `CompositeIndexingStream` class you can pass `indexingSource: CompositeViewIndexingSource, remoteIndexingSource: CompositeViewIndexingSource` and all this block will be like before:
```scala
            case s: ProjectSource       => indexingSource(view.projectRef, progress.offset, s.resourceTag)
            case s: CrossProjectSource  => indexingSource(s.project, progress.offset, s.resourceTag)
            case _: RemoteProjectSource => remoteIndexingSource(view.projectRef, progress.offset, None)
```

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingStream.scala [hunk 11]
```diff
@@ -372,7 +411,8 @@ object CompositeIndexingStream {
       projectsCounts: ProjectsCounts,
       indexingController: CompositeIndexingController,
       projections: Projection[Unit],
-      indexingSource: IndexingSource
+      indexingSource: IndexingSource,
+      metadataPredicates: Set[Node]
```
Comment (umbreak) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingStream.scala:415]: MetadataPredicates

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/IndexingData.scala [hunk 12]
```diff
@@ -1,21 +1,46 @@
...
+  * @param id                    the resource id
+  * @param deprecated            whether the resource is deprecated
+  * @param schema                the resource schema
+  * @param types                 the resource types
   * @param selectPredicatesGraph the graph with the predicates in ''graphPredicates''
   * @param metadataGraph         the graph with the metadata value triples
   * @param source                the original payload of the resource posted by the caller
   */
-final case class IndexingData(selectPredicatesGraph: Graph, metadataGraph: Graph, source: Json)
+final case class IndexingData(
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/IndexingData.scala:23]: This `IndexingData` seems to be the same in blazegraph and elasticsearch. Why don't we move it to `sdk-views` module? The only different I can spot is the `Json`, which can be ignored by using the correct apply method

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala (hunk 1 above)
Comment (wwajerowicz) [reply delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala:45]: When you create `IndexingData` from `ResourceF` it uses `resolvedId`.

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/IndexingData.scala [hunk 13]
```diff
@@ -1,21 +1,46 @@
...
+    metadataGraph: Graph,
+    source: Json
+)
 
 object IndexingData {
   val graphPredicates: Set[Node] = Set(skos.prefLabel, rdfs.label, Vocabulary.schema.name).map(predicate)
+
+  def apply(resource: ResourceF[_], selectPredicatesGraph: Graph, metadataGraph: Graph, source: Json): IndexingData =
+    IndexingData(
+      resource.id,
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/IndexingData.scala:38]: Here it is not using `resource.resolverId`

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala [hunk 14]
```diff
@@ -110,34 +108,19 @@ object BlazegraphIndexingStreamEntry {
...
-                        case (s, p, _) if s == idNode && p == typePredicate => true
-                        case _                                              => false
-                      }
-                      .triples
-                      .map(triple => Iri.unsafe(triple._3.getURI))
-      schema     <- findPredicate(idNode, metagraph, nxv.constrainedBy.iri)
+      types       = graph.rootTypes
+      schema     <- metagraph
+                      .find(id, nxv.constrainedBy.iri)
                       .map(triple => ResourceRef.apply(Iri.unsafe(triple.getURI)))
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/BlazegraphIndexingStreamEntry.scala:118]: A bit cleaner:
```scala
ResourceRef(iri"${triple.getURI}")
```

**Review by umbreak — CHANGES_REQUESTED** [review]

**Review by umbreak — APPROVED** [review]

**Review by bogdanromanx — APPROVED** [review]

**Review by bogdanromanx — APPROVED** [review]

**Review by umbreak — DISMISSED** [review]

**Review by bogdanromanx — APPROVED** [review]

**Pull request merged**